llama_index/indices/query/query_transform/feedback_transform.py

```diff
@@ -1,37 +1,105 @@
+import logging
 from typing import Dict, Optional
 
 from llama_index.evaluation.base import Evaluation
 from llama_index.indices.query.query_transform.base import BaseQueryTransform
 from llama_index.indices.query.schema import QueryBundle
+from llama_index.langchain_helpers.chain_wrapper import LLMPredictor
+from llama_index.llm_predictor.base import BaseLLMPredictor
+from llama_index.prompts.base import Prompt
+from llama_index.response.schema import Response
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_RESYNTHESIS_PROMPT_TMPL = (
+    "Here is the original query:\n"
+    "{query_str}\n"
+    "Here is the response given:\n"
+    "{response}\n"
+    "Here is some feedback from evaluator about the response given.\n"
+    "{feedback}\n"
+    "If you want to resynthesize the query, please return the modified query below.\n"
+    "Otherwise, please return the original query.\n"
+)
+
+DEFAULT_RESYNTHESIS_PROMPT = Prompt(DEFAULT_RESYNTHESIS_PROMPT_TMPL)
 
 
 class FeedbackQueryTransformation(BaseQueryTransform):
     """Transform the query given the evaluation feedback.
 
     Args:
         eval(Evaluation): An evaluation object.
+        llm_predictor(BaseLLMPredictor): An LLM predictor.
+        resynthesize_query(bool): Whether to resynthesize the query.
+        resynthesis_prompt(Prompt): A prompt for resynthesizing the query.
 
     """
 
     def __init__(
         self,
-        evaluation: Evaluation,
+        evaluation: Optional[Evaluation] = None,
+        llm_predictor: Optional[BaseLLMPredictor] = None,
+        resynthesize_query: bool = False,
+        resynthesis_prompt: Optional[Prompt] = None,
     ) -> None:
         super().__init__()
         self.evaluation = evaluation
+        self.llm_predictor = llm_predictor or LLMPredictor()
+        self.should_resynthesize_query = resynthesize_query
+        self.resynthesis_prompt = resynthesis_prompt or DEFAULT_RESYNTHESIS_PROMPT
+
+    def set_eval(self, evaluation: Evaluation) -> None:
```

Comment:
curious: found this quite strange. are evaluations set at different times compare to when the transform is called?

If not, we could just pass the evalutaion as kwargs into the __call__?

Reply:
Yeah... so the reason for this is we want the user to be able to pass their own FeedbackQueryTransformation into the constructor of RetryGuidelineQueryEngine, but as a result we won't have the evaluation at construct time so we need to pass it in before running the feedback transform. We can't pass the eval into self.query_transformer.run(query_bundle) since the interface doesn't support it.

Reply:
Actually, the BaseQueryTransform supports `extra_info` in the interface:

```
    def run(
        self,
        query_bundle_or_str: QueryType,
        extra_info: Optional[Dict] = None,
    ) -> QueryBundle:
```

Reply:
(though I don't particularly like the opaque untyped dict handling)

Reply:
Ah ok I can add it then